pkg/trojan/message.go

```diff
@@ -0,0 +1,278 @@
+	d := c.Data()
+
+	// unmarshal chunk payload into message
+	m := new(Message)
+	// first 40 bytes are span + nonce
+	err := m.UnmarshalBinary(d[40:])
+
+	return m, err
+}
+
+// IsPotential returns true if the given chunk is a potential trojan
+func isPotential(c swarm.Chunk) bool {
+	// chunk must be content-addressed to be trojan
+	if c.Type() != swarm.ContentAddressed {
+		return false
+	}
+
+	data := c.Data()
+	// check for minimum chunk data length
+	trojanChunkMinDataLen := swarm.SpanSize + NonceSize + TopicSize + LengthSize
+	if len(data) < trojanChunkMinDataLen {
+		return false
+	}
+
+	// check for valid trojan message length in bytes #41 and #42
+	messageLen := int(binary.BigEndian.Uint16(data[40:42]))
+	return trojanChunkMinDataLen+messageLen <= len(data)
+}
+
+// checkTargets verifies that the list of given targets is non empty and with elements of matching size
+func checkTargets(targets Targets) error {
+	if len(targets) == 0 {
+		return ErrEmptyTargets
+	}
+	validLen := len(targets[0]) // take first element as allowed length
+	for i := 1; i < len(targets); i++ {
+		if len(targets[i]) != validLen {
+			return ErrVarLenTargets
+		}
+	}
+	return nil
+}
+
+// toChunk finds a nonce so that when the given trojan chunk fields are hashed, the result will fall in the neighbourhood of one of the given targets
+// this is done by iteratively enumerating different nonces until the BMT hash of the serialization of the trojan chunk fields results in a chunk address that has one of the targets as its prefix
+// the function returns a new chunk, with the found matching hash to be used as its address,
+// and its data set to the serialization of the trojan chunk fields which correctly hash into the matching address
+func (m *Message) toChunk(targets Targets, span []byte) (swarm.Chunk, error) {
+	// start out with random nonce
+	nonce := make([]byte, NonceSize)
+	if _, err := rand.Read(nonce); err != nil {
+		return nil, err
+	}
+	nonceInt := new(big.Int).SetBytes(nonce)
+	targetsLen := len(targets[0])
+
+	// serialize message
+	b, err := m.MarshalBinary() // TODO: this should be encrypted
+	if err != nil {
+		return nil, err
+	}
+
+	// hash chunk fields with different nonces until an acceptable one is found
+	for start := time.Now(); ; {
+		s := append(append(span, nonce...), b...) // serialize chunk fields
+		hash1, err := hashBytes(s)
+		if err != nil {
+			return nil, err
+		}
+
+		// take as much of the hash as the targets are long
+		if contains(targets, hash1[:targetsLen]) {
+			// if nonce found, stop loop and return chunk
+			return swarm.NewChunk(swarm.NewAddress(hash1), s), nil
+		}
+		// else, add 1 to nonce and try again
+		nonceInt.Add(nonceInt, big.NewInt(1))
+		// loop around in case of overflow after 256 bits
+		if nonceInt.BitLen() > (NonceSize * swarm.SpanSize) {
+			// Test if timeout after after every 256 iteration
+			if time.Since(start) > (MinerTimeout * time.Second) {
+				break
+			}
```

Comment:
`break`ing after a time-out should not be inside this `if`.

it should be evaluated in every mining loop, not just in the case of having to loop around the 256 bits.

Reply:
well.. checking timeout on every operation is not optimized. instead checking it every X mining operation is good. and i chose 256. To be precise i will check how much time it takes for 256 operations in my laptop and see how it

Reply:
i don't entirely agree, but we can talk about this again when we tackle #444 

it's ok as it is for now :+1: